### PR TITLE
Viz trajectories

### PR DESCRIPTION
**Pull request by JCharline**
# Patient trajectories vizualisation

## Description

PR addressing issue #30.

Function to plot individual event sequences. Events can be one-time/continuous, categorized in event families or relative from an index date. 

Vizualisation parameters are also accepted, such as event colors, labels, and figure parameters.

## Checklist

- [x] If this PR is a bug fix, the bug is documented in the test suite.
- [ ] Changes were documented in the changelog (pending section).
- [x] If necessary, changes were made to the documentation (eg new pipeline).

### CODE REVIEW

File: eds_scikit/plot/event_sequences.py [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+    # Ordering events
+    if family_col is not None:
+        if family_to_index is None:
+            family_to_index = {
+                k: v for v, k in enumerate(list(data_plot[family_col].unique()))
+            }
+
+        data_plot["dim_id"] = data_plot[family_col].map(family_to_index)
+    else:
+        data_plot["dim_id"] = data_plot[event_col].map(
```
Comment (Vincent-Maladiere) [inline eds_scikit/plot/event_sequences.py:120]: Maybe this would be easier to read:
```python
uniques, classes = np.unique(data_plot[event_col], return_inverse=True)
data_plot["dim_id"] = classes
```

File: eds_scikit/plot/event_sequences.py [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+            {k: v for v, k in enumerate(list(data_plot[event_col].unique()))}
+        )
+
+    # Mapping events towards colors and labels
+    if dim_mapping is not None:
+        data_plot["dim_label"] = data_plot[event_col].apply(
+            lambda x: dim_mapping[x]["label"]
+        )
+
```
Comment (Vincent-Maladiere) [inline eds_scikit/plot/event_sequences.py:129]:
```suggestion
```

File: mkdocs.yml [hunk 3]
```diff
@@ -30,38 +30,39 @@ theme:
...
-      - Emergency Units: functionalities/patients-course/is_emergency.md
-      - Visit merging: functionalities/patients-course/visit_merging.md
-      - Consultation dates: functionalities/patients-course/consultation_dates.md
-    - Biology:
-      - functionalities/biology/index.md
-      - functionalities/biology/cleaning.md
-      - functionalities/biology/visualization.md
-      - functionalities/biology/vocabulary.md
-      - functionalities/biology/tutorial.ipynb
+      - Connectors: functionalities/generic/io.ipynb
```
Comment (Vincent-Maladiere) [inline mkdocs.yml:36]: I'm not familiar enough with mkdocs; why were these changes necessary?

**Review by Vincent-Maladiere — COMMENTED** [review]
Hey @JCharline, thanks for the PR! Here is the first round of reviews.
Please format your code using pre-commit like @Thomzoy mentioned to pass the tests.

File: eds_scikit/plot/event_sequences.py [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+    """
+    Plots individual sequences from an events DataFrame. Each event must be recorded with a start date, a name and a `person_id`.
+    Events can be both one-time (only start date given) or longitudinal (both start and end dates).
+    Events can also be aggregated in families using the `family_col` argument.
+    Finally, events labelling and colors can be manually set by providing a `dim_mapping` dictionary.
+
+    Parameters
+    ----------
+    df_events: pd.DataFrame
+        DataFrame gathering the events information. Must contain at least `person_id`, event, t_start and t_end columns.
```
[the remark this answers is not in the document]
Comment (JCharline) [reply eds_scikit/plot/event_sequences.py:35]: Sure, good idea! Thanks for the advice.

File: eds_scikit/plot/event_sequences.py [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+        Column name of the index date to compute relative datetimes for events. For example, it could be the date of inclusion for each patient.
+    family_col: Optional[str] = None
+        Column name of family events. Events of a given family will be plot on the same row.
+    family_to_index: Optional[Dict[str,int]] = None
+        Dictionary mapping family names towards ordering indices.
+    list_person_ids: Optional[List[str]] = None
+        List of person_ids to plot. If None given, all individual sequences will be plot.
+    same_x_axis_scale: Optional[bool] = False
+        Whether to use the same axis scale for all sequences.
+    indiv_height: Optional[int] = 200
```
[the remark this answers is not in the document]
Comment (JCharline) [reply eds_scikit/plot/event_sequences.py:60]: You are right, "subplot_height" seems clearer.

File: eds_scikit/plot/event_sequences.py (hunk 1 above)
Comment (JCharline) [reply eds_scikit/plot/event_sequences.py:120]: I agree, thanks !

File: eds_scikit/plot/event_sequences.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (JCharline) [reply eds_scikit/plot/event_sequences.py:35]: Actually when I skip rows it breaks the docstrings view with mkdocs : 
image

I should maybe leave it as it was, to favour readability in mkdocs. What do you think ?

File: eds_scikit/plot/event_sequences.py [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+                "event_2" : {"color":(200,255,150), "label":"Event 2"}
+            }
+        ```
+    index_date_col: Optional[str] = None
+        Column name of the index date to compute relative datetimes for events. For example, it could be the date of inclusion for each patient.
+    family_col: Optional[str] = None
+        Column name of family events. Events of a given family will be plot on the same row.
+    family_to_index: Optional[Dict[str,int]] = None
+        Dictionary mapping family names towards ordering indices.
+    list_person_ids: Optional[List[str]] = None
```
[the remark this answers is not in the document]
Comment (JCharline) [reply eds_scikit/plot/event_sequences.py:57]: You are right, the default use may crash the notebook. I will add the selection of the 3 first `person_id` if the `list_person_ids=None`, thanks for the suggestion !

File: mkdocs.yml (hunk 3 above)
Comment (JCharline) [reply mkdocs.yml:36]: I think it is a default pep8 formatting, it did it automatically on my VSCode when I opened/saved the file. But it does not seem to be mandatory for pre-commit, so I can just restore the older version.

File: eds_scikit/plot/event_sequences.py [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+        Chart title.
+    seed: int = 0
+        Seed to randomly draw colors when not provided.
+
+    Returns
+    -------
+    chart: alt.Chart
+        Chart with the plotted individual event sequences.
+    """
+    random.seed(seed)
```
Comment (Vincent-Maladiere) [inline eds_scikit/plot/event_sequences.py:79]: While you made a good point about controlling the color randomness, setting the default random seed might lead to weird side effects for the user. Instead, I suggest using a numpy random state to limit the seed to the scope of the function:
```python
rng = np.random.RandomState(seed)
```
and then
```python
colors = [
    tuple(rng.randint(0, 255, size=3))
    for _ in labels
]

```

File: eds_scikit/plot/event_sequences.py [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+
```
Comment (Vincent-Maladiere) [inline eds_scikit/plot/event_sequences.py:80]: It might be super helpful to raise errors upfront when some columns are missing in the Dataframe. `check_columns` is probably the easiest way to do so.

https://github.com/aphp/eds-scikit/blob/c15119c91480dc2faed108b966bb0e66b5fbb597/eds_scikit/utils/checks.py#L192

File: mkdocs.yml (hunk 3 above)
Comment (Vincent-Maladiere) [reply mkdocs.yml:36]: Yes, it's usually better to limit changes that are not directly related to the focus of a PR

File: tests/test_event_sequences.py [hunk 9]
```diff
@@ -0,0 +1,11 @@
+import altair as alt
+
+from eds_scikit.datasets import load_event_sequences
+from eds_scikit.plot.event_sequences import plot_event_sequences
+
+df_events = load_event_sequences()
+
+
+def test_visit_merging():
```
Comment (Vincent-Maladiere) [inline tests/test_event_sequences.py:9]: We are almost done with this review! :) A last thing to focus on is to extend the test to the if / else of the function to have good code coverage. You should set the optional arguments and perform the same check you wrote.

File: eds_scikit/plot/event_sequences.py (hunk 7 above)
Comment (JCharline) [reply eds_scikit/plot/event_sequences.py:79]: Alright, thanks for the suggestion !

File: eds_scikit/plot/event_sequences.py (hunk 8 above)
Comment (JCharline) [reply eds_scikit/plot/event_sequences.py:97]: You are right thanks, just added !

File: tests/test_event_sequences.py (hunk 9 above)
Comment (JCharline) [reply tests/test_event_sequences.py]: It should be better now ! How/Where can I check the % of code coverage for this PR ?

Comment (Vincent-Maladiere) [reply tests/test_event_sequences.py]: This is done automatically in the CI by codecov when this draft becomes a PR, you can also run codecov locally with `pytest --cov=eds_scikit`

File: eds_scikit/plot/event_sequences.py [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+    # Pre-selection of the sequences to plot and required columns.
+    if list_person_ids is None:
+        list_person_ids = list(df_events.person_id.unique()[:3])
+
+    order = {val: idx for idx, val in enumerate(list_person_ids)}
+    df_events = df_events.query("person_id in @list_person_ids").sort_values(
+        by="person_id", key=lambda x: x.map(order)
+    )
+
+    data_plot = df_events.sort_values(by=["person_id", event_start_datetime_col])
```
Comment (Vincent-Maladiere) [inline eds_scikit/plot/event_sequences.py:107]: Could you explain why we need a copy of `df_events` here? Also, we sort `data_plot` by `person_id` without using the order, and I wonder if this is the intended behavior since it can break the ordering.

In case you don't need a copy and if what you want is sorting by `person_id` with a custom order and then sorting with `event_start` I can suggest this:
```python
_, classes = np.unique(df_events["person_id"], return_inverse=True)
df_events["person_id_ordered"] = classes
df_events = (
    df_events.query("person_id in @list_person_ids")
    .sort_values(["person_id_ordered", event_start_datetime_col])
)
```

**Review by Vincent-Maladiere — COMMENTED** [review]
Hi @JCharline, here is probably the last review before merging :)

File: eds_scikit/plot/event_sequences.py [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+        required_columns.append(index_date_col)
+
+    if family_col is not None:
+        required_columns.append(family_col)
+
+    check_columns(df_events, required_columns=required_columns)
+
+    # Pre-selection of the sequences to plot and required columns.
+    if list_person_ids is None:
+        list_person_ids = list(df_events.person_id.unique()[:3])
```
[the remark this answers is not in the document]
Comment (JCharline) [reply eds_scikit/plot/event_sequences.py:100]: Alright, thanks for the suggestion !

File: eds_scikit/plot/event_sequences.py [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+
+    else:
+        x_encoding = f"{event_start_datetime_col}:T"
+        x2_encoding = f"{event_end_datetime_col}:T"
+
+    # Ordering events
+    if family_col is not None:
+        if family_to_index is None:
+            family_to_index = {
+                k: v for v, k in enumerate(list(data_plot[family_col].unique()))
```
[the remark this answers is not in the document]
Comment (JCharline) [reply eds_scikit/plot/event_sequences.py:135]: You are right, thanks !

File: eds_scikit/plot/event_sequences.py (hunk 10 above)
Comment (JCharline) [reply eds_scikit/plot/event_sequences.py:107]: I used to keep a copy of `df_events` in the development process as I returned the `data_plot` for quality checks, but you are raising a good point,it is not required anymore ! I will apply your suggestion, thanks.

Comment (JCharline) [reply eds_scikit/plot/event_sequences.py:107]: I think we should leave the same order as `list_person_id` though, to let the user the opportunity to provide the order they wish.

Comment (JCharline) [reply eds_scikit/plot/event_sequences.py:107]: Actually I think the sorting is not even required... I can just remove it.

Comment (Vincent-Maladiere) [reply eds_scikit/plot/event_sequences.py]: That looks great!

**Review by Vincent-Maladiere — APPROVED** [review]
Thank you for this PR @JCharline, LGTM :)

**Comment by JCharline** [comment]
Thanks @Vincent-Maladiere for the review ! Happy to merge 🥳

**Comment by Thomzoy** [comment]
Very nice PR, thanks @JCharline for the feature and @Vincent-Maladiere for the review ;)
One last thing, could you rebase main and add your contribution to the changelog ?

**Comment by JCharline** [comment]
Thanks @Thomzoy, it should be okay by now !
